### PR TITLE
Secure credential storage

### PR DESCRIPTION
**Pull request by srsanborn**
Move API Client Secret to secure Splunk storage/passwords endpoint.
Shuffle around things with existing config.
Introduce agari_helpers.py module with common code for getting/setting/validating config:
- Common method (REST endpoints vs config file parsing)
- Some groundwork laid for supporting CP and EP apps/input scripts.

### CODE REVIEW

File: splunk/agari/bin/agari_helpers.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        url=url,
+        auth=(agari_settings.get('agari_client_id'), agari_settings.get('agari_client_secret')),
+        verify=False,
+        timeout=10
+    ).json()
+    return True
+
+def update_agari_settings(agari_settings, server_uri, session_key, password_identifier):
+    # Config post:
+    r = requests.post(
```
Comment (djspencer) [inline splunk/agari/bin/agari_helpers.py]: You should probably wrap this in a try with error logging

File: splunk/agari/bin/agari_helpers.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        url=server_uri+'/servicesNS/nobody/' + APP_NAME + '/configs/inputs/agari_config?output_mode=json',
+        data={
+            'param.agari_api_host': agari_settings.get('agari_api_host'),
+            'param.agari_client_id': agari_settings.get('agari_client_id')
+        },
+        headers=splunkd_auth_header(session_key),
+        verify=False).json()
+    # Password post:
+    requests.post(
```
Comment (djspencer) [inline splunk/agari/bin/agari_helpers.py]: Same here - try with error logging

File: splunk/agari/bin/agari_helpers.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    agari_client_secret = None
+    password_url = server_uri + '/servicesNS/nobody/' + APP_NAME + '/storage/passwords/' + password_identifier + '?output_mode=json'
+
+    try:
+        # attempting to retrieve cleartext password, disabling SSL verification for practical reasons
+        result = requests.get(url=password_url, headers=splunkd_auth_header(session_key), verify=False)
+        if result.status_code != 200:
+            print >> sys.stderr, "AgariDebug: ERROR Error: %s" % str(result.json())
+    except Exception, e:
+        print >> sys.stderr, "AgariDebug: ERROR Error sending message: %s" % e
```
Comment (djspencer) [inline splunk/agari/bin/agari_helpers.py]: Would be nice if we could use the common logging methodology.
This probably ends up in the splunkd log ?
Something to think about.

File: splunk/agari/bin/agari_helpers.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    return agari_client_secret
+
+
+# Simple utility methods:
+
+def get_agari_config(server_uri, session_key):
+    # Raw agari_config lookup:
+    # url = server_uri + '/servicesNS/nobody/' + APP_NAME + '/configs/inputs/agari_config?output_mode=json'
+    url = server_uri + '/servicesNS/nobody/agari/configs/inputs/agari_config?output_mode=json'
+    result = requests.get(url=url, headers=splunkd_auth_header(session_key), verify=False)
```
Comment (djspencer) [inline splunk/agari/bin/agari_helpers.py]: Wrap with try and error loggin
OR wrap calling code

File: splunk/agari/bin/agari_input.py [hunk 5]
```diff
@@ -145,54 +151,29 @@ def load_config(config):
...
-    try:
-        for required_config_key in ["api_host", "client_id", "client_secret"]:
-            if not config.has_key(required_config_key):
-                raise Exception, "Missing config value: %s" % (required_config_key)
-
-        if config["api_host"] not in ["https://api.agari.com","https://api-stage.agari.com"]:
-            raise Exception, "Invalid Agari API host specified."
+        # Get credentials from secure storage (as username/password):
+        # read session key sent from splunkd
+        session_key = sys.stdin.readline().strip()
```
Comment (djspencer) [inline splunk/agari/bin/agari_input.py]: Curious how this works.
Reading stdin returns session key ?

File: splunk/agari/bin/agari_input.py [hunk 6]
```diff
@@ -145,54 +151,29 @@ def load_config(config):
...
-            raise Exception, "Agari API host unreachable. Status code returned: %s" % (api_status.status_code)
+        if len(session_key) == 0:
+            sys.stderr.write("Did not receive a session key from splunkd.\n")
+            exit(2)
 
-        access_token = auth()
-        if not access_token:
-            raise Exception, "Agari API authentication error. Check the accuracy of your Agari API credentials."
+        # How do we harden this call?
+        agari_settings = get_agari_settings(splunk.getLocalServerInfo(), session_key, PASSWORD_IDENTIFIER)
```
Comment (djspencer) [inline splunk/agari/bin/agari_input.py]: Maybe check for get_agari_settings == None with error logging.
Probably should exit in this case (or return False)

File: splunk/agari/bin/agari_helpers.py (hunk 3 above)
Comment (srsanborn) [inline splunk/agari/bin/agari_helpers.py]: Yeah, I need to do some investigation into the best Splunk logging mechanism; the `sys.stderr` approach seems to end up in splunkd.log without running the server with the `--debug` flag; but the logging approach we've been using so far seemed to be used in other docs I read. Either way, this'll get unified soon.

File: splunk/agari/bin/agari_input.py (hunk 5 above)
Comment (srsanborn) [inline splunk/agari/bin/agari_input.py]: Yeah, apparently when the script truly runs within a Splunk instance (ie, triggered by its scheduler -- not via `splunk cmd python <filename>`), it receives a session_key in stdin which can be used for various authenticate-y operations, as seen in the code here.

It's "just Splunk magic" insofar as I can tell.

File: splunk/agari/bin/agari_helpers.py (hunk 3 above)
Comment (djspencer) [inline splunk/agari/bin/agari_helpers.py]: Would be nice to keep our separate agari log file if possible.
Happy to work with you to figure out the best approach on this as I have messed around with the logging code a bit.

File: splunk/agari/bin/agari_helpers.py [hunk 7]
```diff
@@ -33,22 +33,37 @@ def validate_agari_settings(agari_settings):
...
+            },
+            headers=splunkd_auth_header(session_key),
+            verify=False).json()
+        if result.status_code != 200:
+            print >> sys.stderr, "AgariDebug: ERROR Error: %s" % str(result.json())
+            return False
+    except Exception, e:
+        print >> sys.stderr, "AgariDebug: ERROR Error sending message: %s" % e
+        return False
+
```
Comment (djspencer) [inline splunk/agari/bin/agari_helpers.py]: There should probably be a 'return True' here for success case.

File: splunk/agari/bin/agari_helpers.py [hunk 8]
```diff
@@ -33,22 +33,37 @@ def validate_agari_settings(agari_settings):
...
+                'password': agari_settings.get('agari_client_secret')
+            },
+            headers=splunkd_auth_header(session_key),
+            verify=False)
+        if result.status_code != 200:
+            print >> sys.stderr, "AgariDebug: ERROR Error: %s" % str(result.json())
+            return False
+    except Exception, e:
+        print >> sys.stderr, "AgariDebug: ERROR Error sending message: %s" % e
+        return False
```
Comment (djspencer) [inline splunk/agari/bin/agari_helpers.py]: Same: need 'return True' case.

File: splunk/agari/bin/agari_input.py [hunk 9]
```diff
@@ -164,8 +164,11 @@ def load_config(config):
             sys.stderr.write("Did not receive a session key from splunkd.\n")
             exit(2)
 
-        # How do we harden this call?
         agari_settings = get_agari_settings(splunk.getLocalServerInfo(), session_key, PASSWORD_IDENTIFIER)
+        if not (agari_settings.has_key('agari_client_id') && agari_settings.has_key('agari_client_secret') && agari_settings.hash_key('agari_api_host')):
+            sys.stderr.write("Did not receive a complete configuration.\n")
```
Comment (djspencer) [inline splunk/agari/bin/agari_input.py]: Can we standardize on a single approach for outputting to stderr?
i.e.
'print >> sys.stderr' OR 'sys.stderr.write'

File: splunk/agari/bin/agari_input.py [hunk 10]
```diff
@@ -165,7 +165,7 @@ def load_config(config):
             exit(2)
 
         agari_settings = get_agari_settings(splunk.getLocalServerInfo(), session_key, PASSWORD_IDENTIFIER)
-        if not (agari_settings.has_key('agari_client_id') && agari_settings.has_key('agari_client_secret') && agari_settings.hash_key('agari_api_host')):
+        if not (agari_settings.has_key('agari_client_id') and agari_settings.has_key('agari_client_secret') and agari_settings.has_key('agari_api_host')):
```
Comment (djspencer) [inline splunk/agari/bin/agari_input.py]: ;-)

**Pull request merged**